### PR TITLE
feat(*): support running iptables with docker 

### LINKED ISSUE #85: Ability to have the entire docker network's traffic redirected to transparent proxy
### Description

Sometimes someone want to run the proxy directly on the host machine and have workloads inside containers.

Have to add to the prerouting chain 2 rules per network:
1. Redirect dns (udp port:53) traffic coming from the interface of the docker network to the sidecar (port 15053)
2. Redirect tcp traffic not going to an ip of this docker network (there's a cidr for each docker network) to the sidecar (port 15001)

```
sudo iptables -t nat -A PREROUTING -i docker0 -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
sudo iptables -t nat -A PREROUTING -i br-+ -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
sudo iptables -t nat -A PREROUTING ! -d 172.17.0.0/16 -i docker0 -p tcp -j REDIRECT --to-ports 15001
sudo iptables -t nat -A PREROUTING ! -d 172.18.0.0/16 -i br-+ -p tcp -j REDIRECT --to-ports 15001
```

It would be nice to have this exposed as a parameter in transparent-proxy like:

```
install-transparent-proxy --vnet <interfaceName>:<cidr>
```

For example in the case above:

```
install-transparent-proxy --vnet docker0:172.17.0.0/16 br-68b3f774cd5c:172.18.0.0/16
```

Info: the docker interface for non default docker network `br-<subsetOfTheId>`. `docker network inspect` is helpful to find the id of a network.

Final result:
```
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:DOCKER - [0:0]
:MESH_INBOUND - [0:0]
:MESH_IN_REDIRECT - [0:0]
:MESH_OUTPUT - [0:0]
:MESH_REDIRECT - [0:0]
-A PREROUTING -m addrtype --dst-type LOCAL -j DOCKER
-A PREROUTING -i docker0 -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
-A PREROUTING -i br-+ -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
-A PREROUTING ! -d 172.17.0.0/16 -i docker0 -p tcp -j REDIRECT --to-ports 15001
-A PREROUTING ! -d 172.18.0.0/16 -i br-+ -p tcp -j REDIRECT --to-ports 15001
-A PREROUTING -p tcp -j MESH_INBOUND
-A OUTPUT -p udp -m udp --dport 53 -m owner --uid-owner 1003 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --uid-owner 102 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 1004 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -m owner --gid-owner 103 -j RETURN
-A OUTPUT -p udp -m udp --dport 53 -j REDIRECT --to-ports 15053
-A OUTPUT ! -d 127.0.0.0/8 -m addrtype --dst-type LOCAL -j DOCKER
-A OUTPUT -p tcp -j MESH_OUTPUT
-A POSTROUTING -s 172.18.0.0/16 ! -o br-68b3f774cd5c -j MASQUERADE
-A POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE
-A POSTROUTING -s 172.17.0.2/32 -d 172.17.0.2/32 -p tcp -m tcp --dport 8000 -j MASQUERADE
-A POSTROUTING -s 172.17.0.2/32 -j MASQUERADE
-A DOCKER -i br-68b3f774cd5c -j RETURN
-A DOCKER -i docker0 -j RETURN
-A DOCKER ! -i docker0 -p tcp -m tcp --dport 8000 -j DNAT --to-destination 172.17.0.2:8000
-A MESH_INBOUND -p tcp -m tcp --dport 15008 -j RETURN
-A MESH_INBOUND -p tcp -m tcp --dport 22 -j RETURN
-A MESH_INBOUND -p tcp -j MESH_IN_REDIRECT
-A MESH_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
-A MESH_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A MESH_OUTPUT ! -d 127.0.0.1/32 -o lo -p tcp -m tcp ! --dport 53 -m owner --uid-owner 1003 -j MESH_IN_REDIRECT
-A MESH_OUTPUT -o lo -p tcp -m tcp ! --dport 53 -m owner ! --uid-owner 1003 -j RETURN
-A MESH_OUTPUT -m owner --uid-owner 1003 -j RETURN
-A MESH_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1004 -j MESH_IN_REDIRECT
-A MESH_OUTPUT -o lo -p tcp -m tcp ! --dport 53 -m owner ! --gid-owner 1004 -j RETURN
-A MESH_OUTPUT -m owner --gid-owner 1004 -j RETURN
-A MESH_OUTPUT -p tcp -m tcp --dport 53 -j REDIRECT --to-ports 15053
-A MESH_OUTPUT -d 127.0.0.1/32 -j RETURN
-A MESH_OUTPUT -j MESH_REDIRECT
-A MESH_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
```

### PR DESCRIPTION
**Pull request by lukidzi**
Added possibility to configure PREROUTING chains, which supports traffic redirection with docker on the host.

Fix: kumahq/kuma-net#85

### CODE REVIEW

File: test/blackbox_tests/consts.go [hunk 1]
```diff
@@ -1,5 +1,5 @@
 package blackbox_tests
 
-const TestCasesAmount = 50
```
[the remark this answers is not in the document]
Comment (slonka) [reply test/blackbox_tests/consts.go:3]: I've hit the same problem and it turned out it's not the case. Does it say `Numerical result out of range` ?

File: test/blackbox_tests/dns_test.go [hunk 2]
```diff
@@ -1027,6 +1028,183 @@ var _ = Describe("Outbound IPv4 DNS/UDP conntrack zone splitting with specific I
 	)
 })
 
+var _ = Describe("Outbound IPv4 DNS/UDP traffic to port 53 from specific input interface", func() {
```
Comment (slonka) [inline test/blackbox_tests/dns_test.go:1031]: We should start thinking about drying up these tests, they're both like ~100 LOC and only like 8 lines are different. Please create an issue for this and add it to the tech backlog

File: test/blackbox_tests/outbound_redirect_test.go [hunk 3]
```diff
@@ -970,3 +970,185 @@ var _ = Describe("Outbound IPv6 TCP traffic to any address:port except ports exc
 		}(),
 	)
 })
+
+var _ = Describe("Outbound IPv4 TCP traffic from specific interface to other ip than excluded", func() {
```
Comment (slonka) [inline test/blackbox_tests/outbound_redirect_test.go:974]: Why do we need this test?

File: test/blackbox_tests/outbound_redirect_test.go [hunk 4]
```diff
@@ -970,3 +970,185 @@ var _ = Describe("Outbound IPv6 TCP traffic to any address:port except ports exc
...
+		Expect(ns2.Cleanup()).To(Succeed())
+	})
+
+	DescribeTable("should be redirected to outbound port",
+		func(serverPort, randomPort uint16) {
+			// given
+			tproxyConfig := config.Config{
+				Redirect: config.Redirect{
+					Outbound: config.TrafficFlow{
+						Enabled: true,
```
Comment (slonka) [inline test/blackbox_tests/outbound_redirect_test.go:1005]: Nothing is excluded on the outbound, why?

File: transparent-proxy/config/config.go [hunk 5]
```diff
@@ -313,7 +326,7 @@ func MergeConfigWithDefaults(cfg Config) Config {
 
 	// .Log
 	result.Log.Enabled = cfg.Log.Enabled
-	if result.Log.Level != DebugLogLevel {
+	if cfg.Log.Level != DebugLogLevel {
```
Comment (slonka) [inline transparent-proxy/config/config.go:329]: is this change needed - it's the same, you just assigned to it right?

File: iptables/builder/builder_nat.go [hunk 6]
```diff
@@ -246,30 +248,71 @@ func addOutputRules(cfg config.Config, dnsServers []string, nat *table.NatTable)
...
 		nat.Prerouting().Append(
 			Jump(Log(PreroutingLogPrefix, cfg.Log.Level)),
 		)
 	}
+
+	if len(cfg.Redirect.VNet.Networks) > 0 {
+		interfaceAndCidr := map[string]string{}
+		for i := 0; i < len(cfg.Redirect.VNet.Networks); i++ {
+			// we accept only first : so in case of IPv6 there should be no problem with parsing
+			pair := strings.SplitN(cfg.Redirect.VNet.Networks[i], ":", 2)
```
Comment (slonka) [inline iptables/builder/builder_nat.go:263]: Can't we just have a pair in the structure? Why do we need to parse this here? with IPv6 having `:` as a delimiter it's a bit confusing that we have another `:` in front of it.

**Review by slonka — COMMENTED** [review]
Left a couple of comments just to understand more.

File: test/blackbox_tests/consts.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukidzi) [reply test/blackbox_tests/consts.go:3]: Yes, and also fd not available

File: iptables/parameters/outinterface_test.go [hunk 7]
```diff
@@ -9,142 +9,37 @@ import (
...
 			Expect(got).To(Equal(want))
 		},
-		Entry("CIDR IPv4 localhost",
-			"127.0.0.1/32", false,
-			"127.0.0.1/32",
+		Entry("interface localhost",
+			"localhost", false,
+			"-o localhost",
 		),
-		Entry("CIDR IPv4 localhost - verbose",
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply iptables/parameters/outinterface_test.go:23]: Those tests were copy past from other and didn't test the actual thing. I've changed this to test the actual thing

File: transparent-proxy/config/config.go (hunk 5 above)
Comment (lukidzi) [reply transparent-proxy/config/config.go:329]: It's needed if you configure it and want to customize it.

File: test/blackbox_tests/outbound_redirect_test.go (hunk 4 above)
Comment (lukidzi) [reply test/blackbox_tests/outbound_redirect_test.go:1005]: It's more about `vnet` configuration. We create rules that requests that is not a interface IP should be redirected

File: test/blackbox_tests/consts.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (slonka) [reply test/blackbox_tests/consts.go:3]: `Numerical result out of range` comes from the label being more than 15 chars in length (https://access.redhat.com/solutions/652593) - we can try changing the suffix generation here https://github.com/lukidzi/kuma-net/blob/78bd4b6514523f0fbe442e48161f467534b13f6b/test/framework/netns/builder.go#L22-L26. Not sure about `fd not available` but that might be connected to the suffix issue.

File: test/blackbox_tests/dns_test.go (hunk 2 above)
Comment (lukidzi) [reply test/blackbox_tests/dns_test.go:1031]: https://github.com/kumahq/kuma-net/issues/104

**Review by bartsmykla — APPROVED** [review]

File: test/blackbox_tests/outbound_redirect_test.go (hunk 3 above)
Comment (lukidzi) [reply test/blackbox_tests/outbound_redirect_test.go:974]: answered below
